Re-introduce the client-example plugin from scratch

zinigor opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->
This is a draft PR that is meant to start a series of changes to the way we consume Jetpack packages. The idea is to start from a package that would use an interface, and then build that interface.

- [ ] Create a plugin skeleton.
- [ ] Discuss and create an example interface usage that will cover the basics of third party Jetpack package consumption.
- [ ] Work out the internals and add new items to this list.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Introduces a new composer repository inside the projects folder.

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
p1HpG7-aRR-p2

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
N/A

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->
* TBA

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
<!-- Guidelines: https://github.com/Automattic/jetpack/blob/master/docs/writing-a-good-changelog-entry.md -->
* TBA

zinigor committed:
[not verified] Created a composer.json file for the client-example plugin.

kbrown9 commented:
If you want boilerplate plugin files, [here's one option](https://github.com/DevinVinson/WordPress-Plugin-Boilerplate).

zinigor committed:
[not verified] Copied a README.md from client-example.

zinigor committed:
[not verified] Added a minimal bootstrap file.

zinigor committed:
[not verified] Added a rough outline of a Configurable trait.

zinigor committed:
Added classes proposed by @sergeymitr.

The original gist is here: https://gist.github.com/sergeymitr/29b9c7d8a98f39087cb9a7cf3cbd5dbd.

zinigor committed:
Added the configurator class for the Connection package.

Based on the work by @sergeymitr in https://gist.github.com/sergeymitr/29b9c7d8a98f39087cb9a7cf3cbd5dbd.

jeherve commented:
Closing this for now because of the lack of activity on this. We can always reopen in the future if needed.
